File: Dsa-day353.py
class Solution:
    def maxActiveSectionsAfterTrade(self, s: str) -> int:
        n = len(s)
        ones = s.count('1')
        if ones == n:
            return n
        one_blocks = []
        i = 0
        while i < n:
            if s[i] == '1':
                j = i
                while j < n and s[j] == '1':
                    j += 1
                left_zero = (i == 0 or s[i - 1] == '0')
                right_zero = (j == n or s[j] == '0')
                if left_zero and right_zero:
                    one_blocks.append((i, j - 1))
                i = j
            else:
                i += 1
        if not one_blocks:
            return ones
        best = ones
        for l, r in one_blocks:
            left_zeros = 0
            left_idx = l - 1
            while left_idx >= 0 and s[left_idx] == '0':
                left_zeros += 1
                left_idx -= 1
            right_zeros = 0
            right_idx = r + 1
            while right_idx < n and s[right_idx] == '0':
                right_zeros += 1
                right_idx += 1
            if left_zeros > 0 and right_zeros > 0:
                best = max(best, ones + left_zeros + right_zeros)
        return min(best, n)

File: Dsa-day353.cpp
class Solution {
public:
    int maxActiveSectionsAfterTrade(string s) {
        int n = s.size();
        int ones = 0;
        for (char c : s) if (c == '1') ones++;
        if (ones == n) return n;
        vector<pair<int, int>> oneBlocks;
        int i = 0;
        while (i < n) {
            if (s[i] == '1') {
                int j = i;
                while (j < n && s[j] == '1') j++;
                bool leftZero = (i == 0 || s[i - 1] == '0');
                bool rightZero = (j == n || s[j] == '0');
                if (leftZero && rightZero) {
                    oneBlocks.push_back({i, j - 1});
                }
                i = j;
            } else {
                i++;
            }
        }
        if (oneBlocks.empty()) return ones;
        int best = ones;
        for (auto &block : oneBlocks) {
            int l = block.first, r = block.second;
            int leftZeros = 0;
            int leftIdx = l - 1;
            while (leftIdx >= 0 && s[leftIdx] == '0') {
                leftZeros++;
                leftIdx--;
            }
            int rightZeros = 0;
            int rightIdx = r + 1;
            while (rightIdx < n && s[rightIdx] == '0') {
                rightZeros++;
                rightIdx++;
            }
            if (leftZeros > 0 && rightZeros > 0) {
                best = max(best, ones + leftZeros + rightZeros);
            }
        }
        return min(best, n);
    }
};